hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/LegacyReplicationManager.java

```diff
@@ -1141,8 +1141,103 @@ private void handleUnderReplicatedHealthy(final ContainerInfo container,
+         */
+        deleteUnhealthyReplicaIfNeeded(container, replicaSet);
+      }
+    }
+  }
+
+  /**
+   * Finds and deletes an unhealthy replica (UNHEALTHY or QUASI_CLOSED) under
+   * certain conditions.
+   */
+  private void deleteUnhealthyReplicaIfNeeded(ContainerInfo container,
+      RatisContainerReplicaCount replicaCount) {
+    LOG.info("Finding an unhealthy replica to delete for container {} with " +
+        "replicas {} to unblock under replication handling.", container,
+        replicaCount.getReplicas());
+
+    int pendingDeletes = getInflightDel(container.containerID());
+    if (pendingDeletes > 0) {
+      LOG.debug("Container {} has {} pending deletes. Will not delete an " +
+          "unhealthy replica for this container.", container, pendingDeletes);
+      return;
+    }
+
+    List<ContainerReplica> replicas = replicaCount.getReplicas();
+    if (replicas.size() < 3) {
+      LOG.debug("Container {} has only {} replicas. Will not delete an " +
+          "unhealthy replica for this container.", container, replicas.size());
+      return;
+    }
+
+    LifeCycleState containerState = container.getState();
+    boolean foundMatchingReplica = false;
+    for (ContainerReplica replica : replicas) {
+      if (compareState(containerState, replica.getState())) {
+        foundMatchingReplica = true;
+        break;
+      }
+    }
+    if (!foundMatchingReplica) {
+      LOG.debug("No matching replica found for container {} with replicas " +
+              "{}. Will not delete any unhealthy replica for this container.",
+          container, replicas);
+      return;
+    }
+
+    List<ContainerReplica> deleteCandidates = new ArrayList<>();
+    // collect unhealthy replicas on in-service, healthy nodes
+    for (ContainerReplica replica : replicas) {
+      try {
+        NodeStatus nodeStatus =
+            nodeManager.getNodeStatus(replica.getDatanodeDetails());
+        if (!nodeStatus.isHealthy() || !nodeStatus.isInService()) {
+          continue;
+        }
+      } catch (NodeNotFoundException e) {
+        LOG.warn("Skipping replica {} when trying to unblock under " +
+            "replication handling.", replica, e);
+        continue;
+      }
+
+      if (replica.getState() == State.QUASI_CLOSED) {
+        deleteCandidates.add(replica);
+      } else if (replica.getState() == State.UNHEALTHY) {
+        deleteCandidates.add(replica);
+      }
+    }
+
+    if (containerState == LifeCycleState.CLOSED) {
+      deleteExcessLowestBcsIDs(container, deleteCandidates, 1);
+      return;
+    }
+
+    // if the container is quasi_closed, delete a replica only if it's seq id
+    // is less, and it doesn't have a unique origin node
+    if (containerState == LifeCycleState.QUASI_CLOSED) {
+      List<ContainerReplica> nonUniqueDeleteCandidates =
+          findNonUniqueDeleteCandidates(replicas, deleteCandidates);
+      deleteExcessLowestBcsIDs(container, nonUniqueDeleteCandidates, 1);
+      return;
```

Comment:
I think this part requires some modification. If we're deleting a quasi closed replica, we need to ensure it's sequence id is actually less than the container. Otherwise we should delete an UNHEALTHY replica.

Reply:
We shouldn't need to delete a quasi-closed replica here. We would get here if we have at least one quasi-closed replica and unhealthy replicas on the rest of the nodes. The quasi-closed replica is under-replicated, which is why we are in `handleUnderReplicatedHealthy`, but it cannot be replicated because the rest of the nodes have unhealthy replicas. So this case should only need to delete unhealthy replicas. I think this would be handled automatically with my suggestion above.

Reply:
I've updated this part to only delete an UNHEALTHY replica when we have a quasi-closed container.